sklearn/preprocessing/_target_encoder.py

```diff
@@ -0,0 +1,214 @@
+import numpy as np
+
+from numbers import Real
+
+from ._encoders import _BaseEncoder
+from ..base import OneToOneFeatureMixin
+from ._target_encoder_fast import _fit_encoding_fast
+from ..utils.validation import _check_y, check_consistent_length
+from ..utils._param_validation import Interval, StrOptions
+
+
+class TargetRegressorEncoder(OneToOneFeatureMixin, _BaseEncoder):
+    """Target Encoder for Regression Targets.
+
+    Each category is encoded based on its effect on the target variable. The encoding
+    scheme mixes the global target mean with the target mean conditioned on the
+    category.
+
+    Read more in the :ref:`User Guide <target_regressor_encoder>`.
+
+    .. note::
+        :class:`TargetRegressorEncoder` uses a cross validation scheme in
+        :meth:`fit_transform` to prevent leaking the target during training. In
+        :meth:`fit_transform`, categorical encodings are obtained from one split and
+        used to encoding the other split. Afterwards, a final categorical encoding is
+        obtained from all the training data, which is used to encode data
+        during :meth:`transform`. This means that `fit().transform()` does not equal
+        `fit_transform()`.
+
+    .. versionadded:: 1.3
+
+    Parameters
+    ----------
+    categories : 'auto' or a list of array-like, default='auto'
+        Categories (unique values) per feature:
+
+        - 'auto' : Determine categories automatically from the training data.
+        - list : ``categories[i]`` holds the categories expected in the ith column. The
+          passed categories should not mix strings and numeric values within a single
+          feature, and should be sorted in case of numeric values.
+
+        The used categories can be found in the ``categories_`` attribute.
+
+    smooth : float, default=30.0
```

Comment:
From the dirty cat implementation and the paper, it seems that you can compute an empirical value (lambda_n / Eq. 8). It would be better to provide an "auto" version if it works.

I am wondering if we should not name this `shrinkage`. I am even wondering if really want to expose the parameter?

Reply:
I kind of like the semantics of `smooth`. For example, lets say there are 40 samples in training and `smooth=10.0`, then we take 20% of the target mean and 80% of the global mean. (This is the same semantics as cuml's TargetEncoder)

In any case, I can get behind an auto option using an empirical value or not exposing it at all.

Reply:
I updated this PR with a `smooth="auto"` that does exactly what the Micci-Barreca paper does (Equation 6). I kept the name "smooth", because it is controlling the `m` parameter from Equation 8 in the paper.